client/engine/messageservice/p2p-message-service/service.go

```diff
@@ -91,26 +83,100 @@ func NewMessageService(ip string, port int, me types.Address, pk []byte) *P2PMes
 	if err != nil {
 		panic(err)
 	}
+
+	mdns := mdns.NewMdnsService(host, "", ms)
+	if err := mdns.Start(); err != nil {
+		panic(err)
+	}
+	ms.mdns = mdns
 	ms.p2pHost = host
 
-	ms.p2pHost.SetStreamHandler(PROTOCOL_ID, func(stream network.Stream) {
-		reader := bufio.NewReader(stream)
-		// Create a buffer stream for non blocking read and write.
-		raw, err := reader.ReadString(DELIMITER)
+	ms.p2pHost.SetStreamHandler(PROTOCOL_ID, ms.msgStreamHandler)
+	ms.p2pHost.SetStreamHandler(PEER_EXCHANGE_PROTOCOL_ID, ms.peerExchangeHandler)
 
-		// An EOF means the stream has been closed by the other side.
-		if errors.Is(err, io.EOF) {
-			stream.Close()
-			return
-		}
-		ms.checkError(err)
-		m, err := protocols.DeserializeMessage(raw)
-		ms.checkError(err)
-		ms.toEngine <- m
+	return ms
+}
+
+// HandlePeerFound is called by the mDNS service when a peer is found.
+func (ms *P2PMessageService) HandlePeerFound(pi peer.AddrInfo) {
+	ms.p2pHost.Peerstore().AddAddr(pi.ID, pi.Addrs[0], peerstore.PermanentAddrTTL)
+	stream, err := ms.p2pHost.NewStream(context.Background(), pi.ID, PEER_EXCHANGE_PROTOCOL_ID)
+	ms.checkError(err)
+	ms.sendPeerInfo(stream)
+	ms.receivePeerInfo(stream)
+	stream.Close()
+}
+
+func (ms *P2PMessageService) msgStreamHandler(stream network.Stream) {
+	reader := bufio.NewReader(stream)
+	// Create a buffer stream for non blocking read and write.
+	raw, err := reader.ReadString(DELIMITER)
+
+	// An EOF means the stream has been closed by the other side.
+	if errors.Is(err, io.EOF) {
 		stream.Close()
+		return
+	}
+	ms.checkError(err)
+	m, err := protocols.DeserializeMessage(raw)
+	ms.checkError(err)
+	ms.toEngine <- m
+	stream.Close()
+}
+
+func (ms *P2PMessageService) peerExchangeHandler(stream network.Stream) {
```

Comment:
Why is this handler needed if we already have HandlePeerFound?

Reply:
There's currently two parts to peer discovery process:
1. MDNS triggers `HandlePeerFound` and a message service is alerted to a new peer ID/multiaddress. We still don't know the SC address of the peer.
2. Whenever we find a a new peer in `HandlePeerFound` we message that peer (using a two-way network stream).  On that stream we send: our peer information (including or SC address)  and then read the peer information from the other message service.

`peerExchangeHandler` is the logic for handling the message received in part 2, and it registered by `ms.p2pHost.SetStreamHandler` in the constructor.

I did find the name slightly confusing so I ended up [embedding](https://github.com/statechannels/go-nitro/blob/23f608b4d04babb0eb5e16891c83327b15f2f499/client/engine/messageservice/p2p-message-service/service.go#L96) the logic into the stream handler